numbat/src/parser.rs

```diff
@@ -392,447 +392,493 @@ impl<'a> Parser<'a> {
-                };
+                    let has_comma = self.match_exact(TokenKind::Comma).is_some();
+                    self.skip_empty_lines();
+                    if self.match_exact(TokenKind::RightParen).is_some() {
+                        break;
+                    }
 
-                if decorator::contains_aliases(&self.decorator_stack) {
+                    if !has_comma && self.peek().kind != TokenKind::RightParen {
+                        return Err(ParseError {
+                                kind: ParseErrorKind::ExpectedCommaEllipsisOrRightParenInFunctionDefinition,
+                                span: self.peek().span,
+                            });
+                    }
+                } else {
                     return Err(ParseError {
-                        kind: ParseErrorKind::AliasUsedOnFunction,
+                        kind: ParseErrorKind::ExpectedParameterNameInFunctionDefinition,
                         span: self.peek().span,
                     });
                 }
+            }
 
-                let mut decorators = vec![];
-                std::mem::swap(&mut decorators, &mut self.decorator_stack);
+            let return_type_annotation = if self.match_exact(TokenKind::Arrow).is_some() {
+                Some(self.type_annotation()?)
+            } else {
+                None
+            };
 
-                Ok(Statement::DefineFunction {
-                    function_name_span,
-                    function_name: fn_name.lexeme.clone(),
-                    type_parameters,
-                    parameters,
-                    body,
-                    return_type_annotation,
-                    decorators,
-                })
+            let (body, local_variables) = if self.match_exact(TokenKind::Equal).is_none() {
+                (None, vec![])
             } else {
-                Err(ParseError {
-                    kind: ParseErrorKind::ExpectedIdentifierAfterFn,
-                    span: self.peek().span,
-                })
-            }
-        } else if self.match_exact(TokenKind::Dimension).is_some() {
-            if let Some(identifier) = self.match_exact(TokenKind::Identifier) {
-                if identifier.lexeme.starts_with("__") {
-                    return Err(ParseError::new(
-                        ParseErrorKind::DoubleUnderscoreTypeNamesReserved,
-                        identifier.span,
-                    ));
-                }
+                self.skip_empty_lines();
+                let body = self.expression()?;
 
-                if self.match_exact(TokenKind::Equal).is_some() {
-                    self.skip_empty_lines();
-                    let mut dexprs = vec![self.dimension_expression()?];
+                // After parsing the `where` statement we have eaten all
+                // the newlines. We're not supposed to do that.
+                // We'll resume the parser to where we were at before doing that.
```

Comment:
> we have eaten all the newlines. We're not supposed to do that.

That is not really the problem though, right? The problem is that we possibly "eat" more than just whitespace. We might have started to parse the beginning of a new statement. For example, in 
```
fn f() = x where x = 1

pi - 3
```
we would continue to parse `pi` as a possible identifier of an additional local variable. But then when we encounter the `-`, we would realize we made a mistake (we would expect a `=` in a variable definition).

I don't know about you, but I'm not super happy with this. The fact that we need to backtrack shows that the grammar is not context free. This will make it harder to write parsers (or "parsers") for Numbat code. For example: editor syntax highlighting integrations.

It's also not ideal that something like
```
fn f(x) = z where z = x * x

not_so_global_variable = 3
```
is parsed without any error, even though it really looks like one. The user probably just forgot to add `let` in front of the second variable definition.

And as you pointed out, this would mean we could not write a streaming parser for Numbat. Something that I initially set out to do, but was always too lazy to finish. It's not a big deal right now, since parsing is quite fast. But it would be another possible downside.

---

Would it make sense to discuss alternative syntax designs again?

Maybe
```
fn f(…) = …
  where
    x = 1,
    y = 2,
    z = 3
```
where the comma would be required between multiple local variable definitions?

Or maybe
```
fn f(…) = …
  where x = 1
    and y = 2
    and z = 3
```

Reply:
Yeah, that definitely seems like a good idea.
I also wanted to make a streaming parser on numbat.

I like that `and` but no strong on my side.
Any preference on your side?

Reply:
Fixed